### DISCUSSION TITLE
Downloading the model for offline use

### DISCUSSION THREAD

**Discussion by daniloimparato** [evolutionaryscale/esm#133] [Q&A]
Hello, everyone.

This is my first time with ESM/Huggingface models in general. I want to load ESM3 on a local server with no internet access. I have tried some general model loading/saving instructions, but no luck so far [[1](https://stackoverflow.com/questions/62261602/downloading-transformers-models-to-use-offline), [2](https://discuss.huggingface.co/t/download-models-for-local-loading/1963), [3](https://huggingface.co/docs/hub/en/models-downloading)]. I have even tried copying my local cache to the server (`~/.cache/huggingface/hub/models--EvolutionaryScale--esm3-sm-open-v1/`). It seems ESM3 works a bit... differently? Would appreciate some guidance.

In other words, how can I change this line from the README to work with a local dir?

```python
model: ESM3InferenceClient = ESM3.from_pretrained("esm3-open").to("cuda")
```

I will keep banging my head at the keyboard, but figured it wouldn't hurt asking this question here as well.

**Comment by PraljakReps** [comment]
One thing you can do is download model and save it on your server with internet server first...

```python
from esm.models.esm3 import ESM3
import torch

client = ESM3.from_pretrained('esm3_sm_open_v1') # download model on local cache
torch.save( client, './esm3_sm_open_v1_full.pth')" # save model onto hard drive
```

Then on your server with no internet connection, do the following:

```python
import torch

client = torch.load('./esm3_sm_open_v1_full.pth') # load model from hard drive

#######
# TEST #
#######

from esm.models.esm3 import ESM3
from esm.sdk.api import ESMProtein, SamplingConfig
from esm.utils.constants.models import ESM3_OPEN_SMALL

# Peptidase S1A, chymotrypsin family: https://www.ebi.ac.uk/interpro/structure/PDB/1utn/
protein = ESMProtein(
    sequence=(
        "FIFLALLGAAVAFPVDDDDKIVGGYTCGANTVPYQVSLNSGYHFCGGSLINSQWVVSAAHCYKSGIQVRLGEDNINVVEG"
        "NEQFISASKSIVHPSYNSNTLNNDIMLIKLKSAASLNSRVASISLPTSCASAGTQCLISGWGNTKSSGTSYPDVLKCLKAP"
        "ILSDSSCKSAYPGQITSNMFCAGYLEGGKDSCQGDSGGPVVCSGKLQGIVSWGSGCAQKNKPGVYTKVCNYVSWIKQTIASN"
    )
)
protein_tensor = client.encode(protein)

output = client.forward_and_sample(
    protein_tensor, SamplingConfig(return_per_residue_embeddings=True)
)
print(output.per_residue_embedding.shape)
```

**Reply by daniloimparato** [reply]
Life saver! Thank you very much, @PraljakReps. I can't try it right now, but no reason to think it wouldn't work.

**Reply by damian238** [reply]
I ran the above programs (cells) and got torch.Size([245, 1536])
It looks like I'm doing something right. I just don't know what to do next.
